src/ngraph/node_input.hpp

```diff
@@ -0,0 +1,95 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/descriptor/input.hpp"
+#include "ngraph/node.hpp"
+
+namespace ngraph
+{
+    class NodeOutput;
+
+    /// \brief A handle for one of a node's inputs.
+    class NodeInput
+    {
+    public:
+        /// \brief Constructs a NodeInput.
+        /// \param node Pointer to the node for the input handle.
+        /// \param index The index of the input.
+        NodeInput(Node* node, size_t index)
+            : m_node(node)
+            , m_index(index)
+        {
+        }
+
+        /// \brief Constructs a NodeInput.
+        /// \param node Shared pointer to the node for the input handle.
+        /// \param index The index of the input.
+        NodeInput(std::shared_ptr<Node> node, size_t index)
```

Comment:
How about we just get rid of this signature? It is easy enough to call the other signature. I'm on a campaign to stomp out shared_ptr everywhere it is not required.

Reply:
I'd be fine with that!

Reply:
Done